text/0000-app-image-extensions.md

```diff
@@ -0,0 +1,182 @@
+
+Note that the `[builder]` section is mutually exclusive with `[stack]`.
+
+### Building an App with Additional Packages
+
+`pack build sclevine/myapp`
+
+project.toml:
+```toml
+...
+[run.extend] # run-metadata-toml-file (stored on new run image metadata)
+packages = ["git"]
+[build.extend] # build-metadata-toml-file (stored on new builder image metadata)
+packages = ["git"]
+...
+```
+
+Behavior: creates a version of the current builder with additional packages and an ephemeral run image with additional packages, then does a normal `pack build`.
+
+Question: should we store the new builder image to make rebuild faster? If so, where? Should we generate a tag for it?
+
+### Upgrade an Extended Builder
+
+Both build and run images: `pack upgrade sclevine/builder`
+
+Run images (including mirrors): `pack upgrade sclevine/run`
+
+### Upgrade an App
+
+`pack upgrade sclevine/myapp`
+
+This will run `pack upgrade sclevine/run`, generate an ephemeral image, and then do the equivalent of `pack rebase sclevine/myapp`.
+
+Attempting to rebase an app directly after it's been upgraded is not permitted and will fail, because the current base is now ephemeral.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Specification
+
+Platforms may extend app images using the above interface by running the `extend` executable as root in a new container and creating an image from the result.
+Extending an image should generate a single layer.
+This should happen prior to the normal CNB build or rebase process.
+The `*-metadata-toml-file` file is used to determine what packages to install.
+The `pkg-cache-dir` directory is used to cache package databases (e.g. to speed up `apt-get update`).
+
+## UX
+
+`pack upgrade` (on base image)
+1. See if new base image available, if so proceed to (3). Otherwise, (2).
+2. Run status, if updates available, proceed to (3). Otherwise, stop.
+3. Run extend and publish new image. Store extension metadata on images. 
+
+If running `pack upgrade` on a builder, perform operation against both builder and run images and publish both.
+If running `pack upgrade` on an app, perform operation against run image (with ephemeral output into app repo), then `pack rebase`.  
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Requires kaniko (or a similar tool) when a docker daemon is unavailable
+- Fully rebasing apps with package dependencies becomes many orders of magnitude slower, due to upgrade needs
+
+# Questions
+[questions]: #questions
+
+What happens if you try to extend an extended builder?
+
+Three options:
+
+1. Fail
+
+2. Treat the extended builder as a normal builder, and override the metadata.
+   Advantages: Allows controlled, multi-level distribution model for changes. Easy to implement.
+   Disadvantages: Requires `pack upgrade`ing multiple levels of builders to fully patch the most extended builder. 
```

Comment:
This seems like the ideal and most clear choice IMO. How do you know, or why should you care if the builder you are extending is extending another builder?

Reply:
If we treat the extending an non-extended builder the same as extending an extended builder, it becomes possible to run `pack upgrade` and still have outdated packages with CVEs. We end up with the same problem that the Dockerfile model has (even though we have an easy fix with (3)).